An online retail business is looking to optimize sales and cut down on costs due to the economic down turn. 

Craft a problem Statement:

1. Problem Statement- In view of the rising inflation, currency devaluation and hike in petroleum products, Eke, an indigenous e-commerce company is exploring commercial strategies to help the company reduce cost of operations, increase sales and revenue to drive their profitability and sustainability goals. 

2. Frame the problem with respect to how your expertise can help them solve the problem-
As a Technology Lawyer and data analyst, I will apply qualitative and Quantitative analytical skills to this project. we will review the current sales and operational strageies, then explore the challenges impacted by inflation, currency devaluation and cost operations. By leveraging data, I aim to identify key areas for cost optimization, revenue generation, talent retention,and ensure commercial strategies that aligns with the overall objectives of the company.

3. Write down steps you would take to arrive at your solutions:

   a. Review current operational team with a view to measure team efficiency- Remote Vs In-person efficiency
   b. Monitor Sales Trend- Income and Expenses
   c. Review Costing Strategies and implement monitoring- Cost-based pricing or Value-based pricing
   d. Review the cost of customer acquisition- How much do we invest in sales or marketing and how much return do we receive.
   e. Develop an OKR sheet to measure traction and impact-
   f. Analyse CRM and Supply chain value chain to ensure cost and profit optimization

In [25]:
import pandas as pd

In [29]:
data=pd.read_csv('Online Retail data.csv')

In [31]:
data

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

       InvoiceDate  Time Stamp  Customer ID         Country  
0          12/1/09  7:45:00 AM      13085.0  United Kingdom  
1          12/1/09  7:45:00 AM      13085.0  United Kingdom  
2          12/1/09  7:45:00 AM      13085.0  United Kingdom  
3          12/1/09  7:45:00 AM      13085.0  United Kingdom  
4          12/1/09  7:45:00 AM      13085.0  United Kingdom  
...            ...         ...          ...             ...  
525456     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525457     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525458     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525459     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525460     12/9/10  8:01:00 AM      17530.0  United Kingdom  

[525461 rows x 8 columns]

In [37]:
data.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

  InvoiceDate  Time Stamp  Customer ID         Country  
0     12/1/09  7:45:00 AM      13085.0  United Kingdom  
1     12/1/09  7:45:00 AM      13085.0  United Kingdom  
2     12/1/09  7:45:00 AM      13085.0  United Kingdom  
3     12/1/09  7:45:00 AM      13085.0  United Kingdom  
4     12/1/09  7:45:00 AM      13085.0  United Kingdom  
5     12/1/09  7:45:00 AM      13085.0  United Kingdom  
6     12/1/09  7:45:00 AM      13085.0  United Kingdom  
7     12/1/09  7:45:00 AM      13085.0  United Kingdom  
8     12/1/09  7:46:00 AM      13085.0  United Kingdom  
9     12/1/09  7:46:00 AM      13085.0  United Kingdom

In [39]:
data.tail(10)

Invoice StockCode                         Description  Quantity  \
525451  538171     22748           POPPY'S PLAYHOUSE KITCHEN         2   
525452  538171     22745          POPPY'S PLAYHOUSE BEDROOM          2   
525453  538171     22558     CLOTHES PEGS RETROSPOT PACK 24          4   
525454  538171     21671        RED SPOT CERAMIC DRAWER KNOB         6   
525455  538171     20971    PINK BLUE FELT CRAFT TRINKET BOX         2   
525456  538171     22271                FELTCRAFT DOLL ROSIE         2   
525457  538171     22750        FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751      FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970  PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931              JUMBO STORAGE BAG SUKI         2   

       InvoiceDate  Time Stamp  Customer ID         Country  
525451     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525452     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525453     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525454     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525455     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525456     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525457     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525458     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525459     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525460     12/9/10  8:01:00 AM      17530.0  United Kingdom

In [41]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      525461 non-null  object 
 1   StockCode    525461 non-null  object 
 2   Description  522533 non-null  object 
 3   Quantity     525461 non-null  int64  
 4   InvoiceDate  525461 non-null  object 
 5   Time Stamp   525461 non-null  object 
 6   Customer ID  417534 non-null  float64
 7   Country      525461 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 32.1+ MB


In [43]:
data.shape

(525461, 8)

In [45]:
data.describe()

Quantity    Customer ID
count  525461.000000  417534.000000
mean       10.337667   15360.645478
std       107.424110    1680.811316
min     -9600.000000   12346.000000
25%         1.000000   13983.000000
50%         3.000000   15311.000000
75%        10.000000   16799.000000
max     19152.000000   18287.000000

This dataset presents a summary of the quantity and customer ID analysis. 
There appears to be significant amount of variations within the dataset. The mean(average) is 10 which means that on average each transaction has about 10 items. However, the Standard deviation is quite high at 107 and shows a signifcant distance from the mean of 10. 

There is an anomaly with the minimum quantity. The -9600 shows some insights for further investigation. Are they items that were returned? Are they errors within the dataset? 

In view of the problem presented by the EKE, the e-commerce company, we see that 25% of the transactions have 1 quantity or less. 50% of the transaction accounts for 3 quantity or less and 75% of the transactions had 10 quantity or less. 

If we have these variations above, how then do we have a signifcant outlier of a transation with 19152?  Which product is this? Who ordered it? Does this product require segmentation?

with about 4383 unique customers, there is definitely something to explore in customer retention. How can we increase the frequency of transaction for the customers who have just one purchase or are less frequent than the others. There are 4681 unique purchases. 3549 of items sold are "white hanging T-light holder". This appears to be the most popular item.

In [49]:
data['Customer ID'].value_counts()

Customer ID
14911.0    5710
17841.0    5114
14606.0    3927
14156.0    2710
12748.0    2665
           ... 
18174.0       1
14827.0       1
17645.0       1
16443.0       1
13222.0       1
Name: count, Length: 4383, dtype: int64

In [51]:
data['Invoice'].value_counts()

Invoice
537434     675
538071     652
537638     601
537237     597
536876     593
          ... 
523981       1
523984       1
523985       1
C523988      1
C516462      1
Name: count, Length: 28816, dtype: int64

In [53]:
data['StockCode'].value_counts()

StockCode
85123A    3516
22423     2221
85099B    2057
21212     1933
21232     1843
          ... 
90005B       1
90197C       1
90080        1
20904        1
21120        1
Name: count, Length: 4632, dtype: int64

In [55]:
data['Description'].value_counts()

Description
WHITE HANGING HEART T-LIGHT HOLDER     3549
REGENCY CAKESTAND 3 TIER               2212
STRAWBERRY CERAMIC TRINKET BOX         1843
PACK OF 72 RETRO SPOT CAKE CASES       1466
ASSORTED COLOUR BIRD ORNAMENT          1457
                                       ... 
stock credited from royal yacht inc       1
VINTAGE METAL CAKE STAND CREAM            1
BLUE BAROQUE FLOCK CANDLE HOLDER          1
S/4 HEART CRYSTAL FRIDGE MAGNETS          1
dotcom email                              1
Name: count, Length: 4681, dtype: int64

In [59]:
data['Quantity'].value_counts()

Quantity
 1       146118
 2        78131
 12       60682
 6        44431
 3        35517
          ...  
-1120         1
 2651         1
 528          1
 681          1
-177          1
Name: count, Length: 825, dtype: int64

In [61]:
data['Country'].value_counts

<bound method IndexOpsMixin.value_counts of 0         United Kingdom
1         United Kingdom
2         United Kingdom
3         United Kingdom
4         United Kingdom
               ...      
525456    United Kingdom
525457    United Kingdom
525458    United Kingdom
525459    United Kingdom
525460    United Kingdom
Name: Country, Length: 525461, dtype: object>

In [63]:
data.isnull()

Invoice  StockCode  Description  Quantity  InvoiceDate  Time Stamp  \
0         False      False        False     False        False       False   
1         False      False        False     False        False       False   
2         False      False        False     False        False       False   
3         False      False        False     False        False       False   
4         False      False        False     False        False       False   
...         ...        ...          ...       ...          ...         ...   
525456    False      False        False     False        False       False   
525457    False      False        False     False        False       False   
525458    False      False        False     False        False       False   
525459    False      False        False     False        False       False   
525460    False      False        False     False        False       False   

        Customer ID  Country  
0             False    False  
1             False    False  
2             False    False  
3             False    False  
4             False    False  
...             ...      ...  
525456        False    False  
525457        False    False  
525458        False    False  
525459        False    False  
525460        False    False  

[525461 rows x 8 columns]

In [65]:
data.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Time Stamp          0
Customer ID    107927
Country             0
dtype: int64

In [67]:
data.duplicated().sum()

6979

In [69]:
data_copy = data.copy()

In [71]:
data.dropna(axis=0, inplace =True, how='all')

In [73]:
data.dropna(axis=1, inplace = True, how='all')

In [75]:
data.drop_duplicates()

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

       InvoiceDate  Time Stamp  Customer ID         Country  
0          12/1/09  7:45:00 AM      13085.0  United Kingdom  
1          12/1/09  7:45:00 AM      13085.0  United Kingdom  
2          12/1/09  7:45:00 AM      13085.0  United Kingdom  
3          12/1/09  7:45:00 AM      13085.0  United Kingdom  
4          12/1/09  7:45:00 AM      13085.0  United Kingdom  
...            ...         ...          ...             ...  
525456     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525457     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525458     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525459     12/9/10  8:01:00 AM      17530.0  United Kingdom  
525460     12/9/10  8:01:00 AM      17530.0  United Kingdom  

[518482 rows x 8 columns]

In [77]:
data.to_csv('cleaned assignment data.csv')